pydantic/_internal/_core_utils.py

```diff
@@ -157,11 +163,21 @@ def _record_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.Cor
 
 def collect_invalid_schemas(schema: core_schema.CoreSchema) -> list[core_schema.CoreSchema]:
     invalid_schemas: list[core_schema.CoreSchema] = []
+    child_invalid = False  # track recursion
 
     def _is_schema_valid(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.CoreSchema:
-        if s.get('metadata', {}).get('invalid'):
+        nonlocal child_invalid
+        metadata = s.setdefault('metadata', {})
+        invalid = metadata.get('invalid', None)
+        if invalid is True:
             invalid_schemas.append(s)
-        return recurse(s, _is_schema_valid)
+        elif invalid is False:
+            return s
+        s = recurse(s, _is_schema_valid)
+        if invalid is True:
+            child_invalid = True
+        metadata['invalid'] = child_invalid
+        return s
```

Comment:
Can you elaborate on the motivation of this change? Maybe needs a comment, it's quite winding logic.

Reply:
The `invalid` thing already existed. What this is doing is marking a schema as "there are no invalid schemas in this subtree" (which is the most common case) so that we can skip recursing into it after the first time we recurse through it.

Reply:
I'll wait to make that change until #7529 is merged so I can re-use `post_process_schema` and such